### DISCUSSION TITLE
Codec Priority

### DISCUSSION THREAD

**Discussion by LynxEric** [sipsorcery-org/sipsorcery#1183] [Q&A]
Hey all, I'm quite new to SIP and am confused about SipSorcery's codec priority handling, as well as codec selection generally. For context, I have a very simple use case of sending audio to endpoints over a SIP call. Supporting G722 seemed like an obvious choice for its greater audio quality, as well as PCMU for low bandwidth usage and its wide amount of support. However, regardless of how the codecs are prioritized on the device and it's VoIP infrastructure, PCMU is always chosen over G722 for RTP. I've included G729 to test this and priority seems to go to G722 regardless of priority configuration on device, PBX, and Trunk.
Does this have to do with how codecs are negotiated with SDP? It seems like the codec listed first gets priority over subsequent codec offerings, but from what I understand the endpoint receiving the audio gets final say in what codec to use.

**Comment by sipsorcery** [comment]
You are correct, the codec selection logic is very unophisticated, it simply uses the [first available compatible codec](https://github.com/sipsorcery-org/sipsorcery/blob/6cf4b724b3f2aac17f4aec54085055eb191508ef/src/app/Media/VoIPMediaSession.cs#L183). It could defnitely do with some improvement.

There is a way to specify you only want certain codecs as below but that's not really the same thing.

```
var audioSession = new WindowsAudioEndPoint(new AudioEncoder());
audioSession.RestrictFormats(x => x.Codec == AudioCodecsEnum.PCMA || x.Codec == AudioCodecsEnum.PCMU);
audioSession.RestrictFormats(x => x.Codec == AudioCodecsEnum.PCMA);
```

**Reply by sipsorcery** [reply]
Actually turns out using the first codec in the SDP response IS the [expected behaviour](https://datatracker.ietf.org/doc/html/rfc3264#section-7). Therefore the improvement needed here is the ability to specify the order of the formats in the SDP offer/answer so that the desired codec is first.
